xtm/xtm_api.cc

```diff
@@ -0,0 +1,364 @@
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+#include "xtm_api.h"
+#include "xtm_scsp_queue.h"
+#include "xtm_config.h"
+
+#include <unistd.h>
+#include <stdint.h>
+#include <assert.h>
+#include <fcntl.h>
+#include <errno.h>
+#ifdef TARANTOOL_XTM_USE_EVENTFD
+#include <sys/eventfd.h>
+#endif /* defined(TARANTOOL_XTM_USE_EVENTFD) */
+
+#define XTM_PIPE_SIZE 4096
+
+union xtm_msg {
+	/**
+	 * Anonymous structure for storing functions and their arguments,
+	 * when using the xtm API in dispatch/invoke pattern.
+	 */
+	struct {
+		xtm_queue_fun_t fun;
+		void *fun_arg;
+	};
+	/**
+	 * Pointer to save message, when using xtm API
+	 * in send/recv pattern.
+	 */
+	void *ptr;
+};
+
+struct xtm_queue {
+	/**
+	 * File descriptor that the consumer thread must poll,
+	 * to know when new messages are added to the queue.
+	 */
+	int consumer_read_fd;
+	/**
+	 * File descriptor to which the producer thread writes
+	 * to notify the consumer thread of new messages in the queue.
+	 */
+	int consumer_write_fd;
+	/**
+	 * File descriptor that the producer thread must poll,
+	 * to know when there is a free space in qeuue.
+	 */
+	int producer_read_fd;
+	/**
+	 * File descriptor to which the consumer thread writes
+	 * to notify the producer thread about free space in queue.
+	 */
+	int producer_write_fd;
+	/**
+	 * Flags, that define queue behaviour.
+	 */
+	unsigned flags;
+	/**
+	 * Flag indicates, that producer couldn't put an item in
+	 * the queue, and is waiting for notification.
+	 */
+	bool is_producer_should_be_notified;
+	/** Message queue, it's size must be power of two */
+	struct xtm_scsp_queue<union xtm_msg> queue;
+};
+
+static inline int
+notify_fd(int fd)
+{
+	static uint64_t tmp = 1;
+	ssize_t cnt;
+	/*
+	 * We must write 8 byte value, because for linux we
+	 * used eventfd, which require to write >= 8 byte at once.
+	 * Also in case of EINTR we retry to write
+	 */
+	while ((cnt = write(fd, &tmp, sizeof(tmp))) < 0 && errno == EINTR)
+		;
+	return ((cnt == sizeof(tmp) || errno == EAGAIN) ? 0 : -1);
+}
+
+static inline int
+create_fds(int *read_fd, int *write_fd)
+{
+	int fds[2];
+	assert(read_fd != NULL);
+	assert(write_fd != NULL);
+
+#ifdef TARANTOOL_XTM_USE_EVENTFD
+	if ((fds[0] = eventfd(0, 0)) < 0)
+#else /* !defined(TARANTOOL_XTM_USE_EVENTFD) */
+	if (pipe(fds) < 0)
+#endif /* defined(TARANTOOL_XTM_USE_EVENTFD) */
+		return -1;
+
+#ifdef TARANTOOL_XTM_USE_EVENTFD
+	*read_fd = *write_fd = fds[0];
+#else /* !defined(TARANTOOL_XTM_USE_EVENTFD) */
+	*read_fd = fds[0];
+	*write_fd = fds[1];
+#endif /* defined(TARANTOOL_XTM_USE_EVENTFD) */
+	return 0;
+}
+
+struct xtm_queue *
+xtm_queue_new(unsigned size, unsigned flags)
+{
+	int save_errno = 0;
+	struct xtm_queue *queue = (struct xtm_queue *)
+		malloc(sizeof(struct xtm_queue) +
+		       size * sizeof(union xtm_msg));
+	if (queue == NULL)
+		return NULL;
+
+	assert((flags & (~XTM_QUEUE_MUST_CLOSE_READFD)) == 0);
```

Comment:
I suggest ```#define```s with ORed flags (even if there is only one) for such asserts, not copy-paste - this is more readable and extensible. There and below.

Reply:
Fixed